Enhance type inference on Runtime Context

gpanakkal opened the pull request:
## Description

This PR permits the generic type on RuntimeContext to be extended via module declaration:
```
export * from '@mastra/core/runtime-context';

declare module '@mastra/core/runtime-context' {
  interface RuntimeContextInterface {
    name: string;
    age: number;
  }
}
```

Per #4467, this improves type inference of values when using keys from `RuntimeContext.keys()` or `RuntimeContext.entries()`:
```
const runtimeContext = new RuntimeContext();

for (const [key, value] of runtimeContext.entries()) {
  if (key === 'age') {
    // value is correctly inferred as number
  } else {
    // value is correctly inferred as string
  }
}
```

## Related Issue(s)

https://github.com/mastra-ai/mastra/issues/4467

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Code refactoring
- [ ] Performance improvement
- [ ] Test update

## Checklist

- [ ] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works

gpanakkal committed:
`RuntimeContext.entries()`: improve type narrowing on return value

Co-authored-by: Morgan Davis <morgansumdavis@gmail.com>

gpanakkal committed:
`RuntimeContext`: add type tests

Co-authored-by: Morgan Davis <morgansumdavis@gmail.com>

gpanakkal committed:
RuntimeContext:
- Relax key types on `.set()` and `.get()` to allow nonexistent keys
- Add interface extension tests

Co-authored-by: Morgan Davis <morgansumdavis@gmail.com>

wardpeet commented:
I've been trying tis branch and the code looks good but the DX is not what I was hoping for. I have to overrdie teh interface to make sure that I have typing in my mastra class. Using it wihtout it means I have no typings.

This is not giving me any types
```
const context = new RuntimeContext();
context.set('yo', 'world')
const x = context.get('yo') 
```

I'm not sure how to fix this properly, though, so I'll do a bit of discovery first

gpanakkal commented:
> I've been trying tis branch and the code looks good but the DX is not what I was hoping for. I have to overrdie teh interface to make sure that I have typing in my mastra class. Using it wihtout it means I have no typings.
> 
> This is not giving me any types
> 
> ```
> const context = new RuntimeContext();
> context.set('yo', 'world')
> const x = context.get('yo') 
> ```
> 
> I'm not sure how to fix this properly, though, so I'll do a bit of discovery first

I was able to do it like this:
```
export interface RuntimeContextInterface {};

export class RuntimeContext<
Values extends Record<string, any> = RuntimeContextInterface
>

set<K extends string, V extends K extends keyof Values ? Values[K] : unknown>(
    key: K, value: V,
  ): asserts this is RuntimeContext<Values & Record<K, V>>
 ```
 but this only works for the first invocation of `.set`, so this test fails on the final line:
 ```
 it('should infer the correct type for added keys', () => {
      const runtimeContext = new RuntimeContext(); 
      runtimeContext.set('new-key', 'value');
      runtimeContext.set('count', 99);
      expectTypeOf(runtimeContext.get('new-key')).toBeString();
      expectTypeOf(runtimeContext.get('count')).toBeNumber(); // infers type as `(key: never) => never`
    });
 ```
Since `Values` is inferred to include `{ new-key: string }` by the time `runtimeContext.set('count', 99)` is called, I expected the type to expand again to include `{ count: number }`, but it doesn't. Is this some limitation of TS itself?

gpanakkal committed:
`RuntimeContext`:
- Fall back on inferring `unknown` when passing unrecognized keys to `.get()`
- Raise an error if passing a value of the wrong type to `.set()`
- Remove some `as` assertions
- Filter to string keys only when using `keyof`
- Refactor type tests and update TS configuration to ensure encapsulation
- Add tests for default type argument

Co-authored-by: Morgan Davis <morgansumdavis@gmail.com>

gpanakkal committed:
Add changeset; add examples to documentation

Co-authored-by: Morgan Davis <morgansumdavis@gmail.com>

DanielSLew committed:
Merge branch 'main' into runtime-context-types

wardpeet commented:
So that's fine, setter won't need to update the types. I do want to be able to do
```
const runtimeContext = new RuntimeContext<{ something: string, else: number }>(); 
```

gpanakkal committed:
Exclude type tests from core builds

Co-authored-by: Morgan Davis <morgansumdavis@gmail.com>

gpanakkal committed:
Typo fix

Co-authored-by: Morgan Davis <morgansumdavis@gmail.com>

gpanakkal committed:
Merge branch 'main' into runtime-context-types

DanielSLew committed:
Merge branch 'main' into runtime-context-types

gpanakkal committed:
Merge branch 'main' into runtime-context-types

gpanakkal committed:
Verify runId type before calling workflow.createRun

Co-authored-by: Morgan Davis <morgansumdavis@gmail.com>

DanielSLew committed:
Merge branch 'main' into runtime-context-types
